In [1]:
import pandas as pd

## Examples of discrepancies in UDS crosswalks across years

In [11]:
df_output = pd.read_csv('../data/output/zip2zcta_master_xwalk/zip2zcta_master_xwalk.csv')
df_uds = pd.read_csv("../data/intermediate/uds_clean_xwalk/uds_clean_xwalk.csv", sep=";")

/tmp/ipykernel_2837785/2870395200.py:2: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  df_uds = pd.read_csv("../data/intermediate/uds_clean_xwalk/uds_clean_xwalk.csv", sep=";")


In [5]:
# Number of distinct zcta mapped to a zip code across years
df_output.zcta_matches.value_counts()

zcta_matches
1    39896
2     1452
3      119
4        8
5        1
Name: count, dtype: int64

In [12]:
# Get the zipcode that is mapped to the most zcta
df_output[df_output.zcta_matches == 5]

zip   zcta  year_match census_edition state  is_area  is_self_mapped  \
30691  72199  72118        2021      ZCTA5CE20    AR     True           False   

       is_post_office                  info  min_match  max_match  \
30691           False  Spatial join to ZCTA       2009       2021   

       year_matches  zcta_matches  
30691            13             5

In [13]:
df_uds[['zip', 'zcta']][df_uds.zip == 72199].drop_duplicates()

zip   zcta
30633   72199  72116
175513  72199  72120
402021  72199  72199
443111  72199  72118
484199  72199  72002

In [14]:
df_uds[df_uds.zip == 72199]

zip   zcta  year state  is_area  is_self_mapped  is_post_office  \
30633   72199  72116  2009    AR    False           False           False   
72156   72199  72116  2010    AR    False           False           False   
113375  72199  72116  2011    AR    False           False           False   
154746  72199  72116  2012    AR    False           False           False   
175513  72199  72120  2013    AR     True           False           False   
216768  72199  72120  2014    AR     True           False           False   
278561  72199  72120  2015    AR     True           False           False   
319748  72199  72120  2016    AR     True           False           False   
360900  72199  72120  2017    AR     True           False           False   
402021  72199  72199  2018    AR     True            True           False   
443111  72199  72118  2019    AR     True           False           False   
484199  72199  72002  2020    AR     True           False           False   
525297  72199  72118  2021    AR     True           False           False   

                        info  
30633                    NaN  
72156                    NaN  
113375                   NaN  
154746                   NaN  
175513                   NaN  
216768                   NaN  
278561                   NaN  
319748  Spatial join to ZCTA  
360900  Spatial join to ZCTA  
402021  Spatial join to ZCTA  
443111  Spatial join to ZCTA  
484199  Spatial join to ZCTA  
525297  Spatial join to ZCTA

Looking at the areas in Google Maps, the five zcta's (72116, 72120, 72199, 72118, 72002) that were assigned to zipcode 72199 are very close to each other. 

## State

In [44]:
nstate = df_uds[['zip', 'zcta', 'year', 'state']].groupby(['zip', 'zcta', 'year']).nunique('state')
nstate.describe()

state
count  535977.000000
mean        0.999582
std         0.020439
min         0.000000
25%         1.000000
50%         1.000000
75%         1.000000
max         1.000000

In [45]:
nstate = df_uds[['zcta', 'year', 'state']].groupby(['zcta', 'year']).nunique('state')
nstate.describe()

state
count  430743.000000
mean        1.000163
std         0.013456
min         1.000000
25%         1.000000
50%         1.000000
75%         1.000000
max         3.000000

In [46]:
nstate[nstate.state > 1]

state
zcta  year       
3904  2009      2
      2010      2
      2011      2
      2012      2
20008 2009      3
...           ...
73949 2021      2
98671 2009      2
      2010      2
      2011      2
      2012      2

[66 rows x 1 columns]

In [52]:
df_uds[df_uds.zcta == 20008]

zip   zcta  year state  is_area  is_self_mapped  is_post_office  \
7657    20008  20008  2009    DC     True            True           False   
8003    20588  20008  2009    MD    False           False           False   
8009    20598  20008  2009    VA    False           False           False   
49021   20008  20008  2010    DC     True            True           False   
49367   20588  20008  2010    MD    False           False           False   
49373   20598  20008  2010    VA    False           False           False   
90397   20008  20008  2011    DC     True            True           False   
90743   20588  20008  2011    MD    False           False           False   
90749   20598  20008  2011    VA    False           False           False   
131768  20008  20008  2012    DC     True            True           False   
132114  20588  20008  2012    MD    False           False           False   
132120  20598  20008  2012    VA    False           False           False   
185408  20008  20008  2013    DC     True            True           False   
226663  20008  20008  2014    DC     True            True           False   
255649  20008  20008  2015    DC     True            True           False   
296885  20008  20008  2016    DC     True            True           False   
338055  20008  20008  2017    DC     True            True           False   
379230  20008  20008  2018    DC     True            True           False   
420310  20008  20008  2019    DC     True            True           False   
461408  20008  20008  2020    DC     True            True           False   
502499  20008  20008  2021    DC     True            True           False   

                    info  
7657                 NaN  
8003                 NaN  
8009                 NaN  
49021                NaN  
49367                NaN  
49373                NaN  
90397                NaN  
90743                NaN  
90749                NaN  
131768               NaN  
132114               NaN  
132120               NaN  
185408               NaN  
226663               NaN  
255649               NaN  
296885  Zip Matches ZCTA  
338055  Zip Matches ZCTA  
379230  Zip Matches ZCTA  
420310  ZIP Matches ZCTA  
461408  Zip matches ZCTA  
502499  Zip matches ZCTA

In [49]:
nstate = df_output[['zcta', 'state']].groupby(['zcta']).nunique('state')
nstate.describe()

state
count  33144.000000
mean       1.000060
std        0.007768
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        2.000000

In [50]:
nstate[nstate.state > 1]

state
zcta        
20015      2
73949      2

In [51]:
df_output[df_output.zcta == 20015]

zip   zcta  year_match census_edition state  is_area  is_self_mapped  \
7682  20015  20015        2021      ZCTA5CE20    DC     True            True   
7708  20046  20015        2014      ZCTA5CE10    MD    False           False   
7709  20047  20015        2021      ZCTA5CE20    DC    False           False   
7735  20076  20015        2021      ZCTA5CE20    DC    False           False   

      is_post_office                  info  min_match  max_match  \
7682           False      Zip matches ZCTA       2009       2021   
7708            True                   NaN       2009       2014   
7709            True  Spatial join to ZCTA       2009       2021   
7735            True  Spatial join to ZCTA       2009       2021   

      year_matches  zcta_matches  
7682            13             1  
7708             6             1  
7709            12             2  
7735            12             2